[Bug][broker] BrokerId npe when broker restart

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Read release policy

- [X] I understand that unsupported versions don't get bug fixes. I will attempt to reproduce the issue on a supported version of Pulsar client and Pulsar broker.


### Version

broker: 3.0.4
client: 2.9.5

### Minimal reproduce step

client continue to produce and consume

broker restart

### What did you expect to see?

no npe 

### What did you see instead?

```
2024-06-25T13:27:33,051+0800 [main] INFO  org.apache.pulsar.broker.service.BrokerService - Started Pulsar Broker service on /0.0.0.0:6651, TLS: (none), listener: (none)
2024-06-25T13:27:33,129+0800 [main] INFO  org.apache.pulsar.broker.service.BrokerService - Started Pulsar Broker service on /0.0.0.0:6751, TLS: OPENSSL, listener: (none)
2024-06-25T13:27:33,153+0800 [main] INFO  org.apache.pulsar.broker.service.BrokerService - Scheduling a thread to check backlog quota after [60] seconds in background
2024-06-25T13:27:33,155+0800 [main] INFO  org.apache.pulsar.broker.service.persistent.DispatchRateLimiter - configured broker message-dispatch rate DispatchRateImpl(dispatchThrottlingRateInMsg=0, dispatchThrottlingRateInByte=0, relativeToPublishRate=false, ratePeriodInSecond=1)
2024-06-25T13:27:33,155+0800 [main] INFO  org.apache.pulsar.broker.service.persistent.DispatchRateLimiter - setting message-dispatch-rate DispatchRateImpl(dispatchThrottlingRateInMsg=0, dispatchThrottlingRateInByte=0, relativeToPublishRate=false, ratePeriodInSecond=1)
2024-06-25T13:27:33,175+0800 [main] INFO  org.eclipse.jetty.util.log - Logging initialized @15239ms to org.eclipse.jetty.util.log.Slf4jLog
2024-06-25T13:27:33,175+0800 [pulsar-io-23-3] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51004] connected with role=admin using authMethod=token, clientVersion=2.9.5, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,175+0800 [pulsar-io-23-6] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51010] connected with role=admin using authMethod=token, clientVersion=2.9.5, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,175+0800 [pulsar-io-23-5] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51008] connected with role=admin using authMethod=token, clientVersion=2.9.4, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,175+0800 [pulsar-io-23-1] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51000] connected with role=admin using authMethod=token, clientVersion=2.9.5, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,175+0800 [pulsar-io-23-4] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51006] connected with role=admin using authMethod=token, clientVersion=2.9.4, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,175+0800 [pulsar-io-23-2] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51002] connected with role=admin using authMethod=token, clientVersion=2.9.5, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,191+0800 [pulsar-io-23-7] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51012] connected with role=admin using authMethod=token, clientVersion=2.9.4, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,202+0800 [pulsar-io-23-4] INFO  org.apache.pulsar.broker.service.ServerCnx - [[id: 0xf8f4511f, L:/xxx:6651 - R:/xxx:51006]] Subscribing on topic xxx-partition-115 / debug_gateway. consumerId: 981
2024-06-25T13:27:33,202+0800 [pulsar-io-23-7] INFO  org.apache.pulsar.broker.service.ServerCnx - [[id: 0xb3980641, L:/xxx:6651 - R:/xxx:51012]] Subscribing on topic xxx-partition-115 / debug_gateway. consumerId: 1148
2024-06-25T13:27:33,202+0800 [pulsar-io-23-5] INFO  org.apache.pulsar.broker.service.ServerCnx - [[id: 0xd26bffc1, L:/xxx:6651 - R:/xxx:51008]] Subscribing on topic xxx-partition-129 / debug_gateway. consumerId: 1113
2024-06-25T13:27:33,242+0800 [pulsar-io-23-9] INFO  org.apache.pulsar.broker.service.ServerCnx - [/xxx:51014] connected with role=admin using authMethod=token, clientVersion=2.9.4, clientProtocolVersion=19, proxyVersion=null
2024-06-25T13:27:33,243+0800 [pulsar-io-23-9] INFO  org.apache.pulsar.broker.service.ServerCnx - [[id: 0x2e7cd8b2, L:/xxx:6651 - R:/xxx:51014]] Subscribing on topic xxx-partition-108 / debug_gateway. consumerId: 897
2024-06-25T13:27:33,378+0800 [metadata-store-29-1] ERROR org.apache.pulsar.broker.service.ServerCnx - [/xxx:51000] Failed to create topic xxx-partition-56, producerId=56

java.util.concurrent.CompletionException: java.lang.NullPointerException: brokerId is not initialized before start has been called
        at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[?:?]
        at org.apache.pulsar.broker.service.BrokerService.lambda$loadOrCreatePersistentTopic$59(BrokerService.java:1607) ~[org.apache.pulsar-pulsar-broker-3.0.4.1.jar:3.0.4.1]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147) ~[?:?]
        at org.apache.pulsar.common.naming.NamespaceBundleFactory.lambda$doLoadBundles$2(NamespaceBundleFactory.java:107) ~[org.apache.pulsar-pulsar-broker-3.0.4.1.jar:3.0.4.1]
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147) ~[?:?]
        at org.apache.pulsar.metadata.impl.ZKMetadataStore.handleGetResult(ZKMetadataStore.java:275) ~[org.apache.pulsar-pulsar-metadata-3.0.4.1.jar:3.0.4.1]
        at org.apache.pulsar.metadata.impl.ZKMetadataStore.lambda$batchOperation$5(ZKMetadataStore.java:219) ~[org.apache.pulsar-pulsar-metadata-3.0.4.1.jar:3.0.4.1]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[io.netty-netty-common-4.1.105.Final.jar:4.1.105.Final]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.NullPointerException: brokerId is not initialized before start has been called
        at java.util.Objects.requireNonNull(Objects.java:233) ~[?:?]
        at org.apache.pulsar.broker.PulsarService.getBrokerId(PulsarService.java:1804) ~[org.apache.pulsar-pulsar-broker-3.0.4.1.jar:3.0.4.1]
        at org.apache.pulsar.broker.service.BrokerService.lambda$checkTopicNsOwnership$100(BrokerService.java:2231) ~[org.apache.pulsar-pulsar-broker-3.0.4.1.jar:3.0.4.1]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150) ~[?:?]
        ... 15 more
```

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java
@@ -291,6 +291,7 @@ public class PulsarService implements AutoCloseable, ShutdownService {
     private final ExecutorProvider transactionExecutorProvider;
     private final DefaultMonotonicSnapshotClock monotonicSnapshotClock;
     private String brokerId;
+    private final CompletableFuture<Void> readyForIncomingRequestsFuture = new CompletableFuture<>();
 
     public enum State {
         Init, Started, Closing, Closed
@@ -999,6 +1000,9 @@ public class PulsarService implements AutoCloseable, ShutdownService {
 
             this.metricsGenerator = new MetricsGenerator(this);
 
+            // the broker is ready to accept incoming requests by Pulsar binary protocol and http/https
+            readyForIncomingRequestsFuture.complete(null);
+
             // Initialize the message protocol handlers.
             // start the protocol handlers only after the broker is ready,
             // so that the protocol handlers can access broker service properly.
@@ -1047,10 +1051,20 @@ public class PulsarService implements AutoCloseable, ShutdownService {
             state = State.Started;
         } catch (Exception e) {
             LOG.error("Failed to start Pulsar service: {}", e.getMessage(), e);
-            throw new PulsarServerException(e);
+            PulsarServerException startException = new PulsarServerException(e);
+            readyForIncomingRequestsFuture.completeExceptionally(startException);
+            throw startException;
         } finally {
             mutex.unlock();
         }
+    }
+
+    public void runWhenReadyForIncomingRequests(Runnable runnable) {
+        readyForIncomingRequestsFuture.thenRun(runnable);
+    }
+
+    public void waitUntilReadyForIncomingRequests() throws ExecutionException, InterruptedException {
+        readyForIncomingRequestsFuture.get();
     }
 
     protected BrokerInterceptor newBrokerInterceptor() throws IOException {

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java
@@ -36,7 +36,6 @@ import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
@@ -167,10 +166,10 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
 
     private TopBundleLoadDataReporter topBundleLoadDataReporter;
 
-    private ScheduledFuture brokerLoadDataReportTask;
-    private ScheduledFuture topBundlesLoadDataReportTask;
+    private volatile ScheduledFuture brokerLoadDataReportTask;
+    private volatile ScheduledFuture topBundlesLoadDataReportTask;
 
-    private ScheduledFuture monitorTask;
+    private volatile ScheduledFuture monitorTask;
     private SplitScheduler splitScheduler;
 
     private UnloadManager unloadManager;
@@ -199,7 +198,7 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
 
     private final ConcurrentHashMap<String, CompletableFuture<Optional<BrokerLookupData>>>
             lookupRequests = new ConcurrentHashMap<>();
-    private final CountDownLatch initWaiter = new CountDownLatch(1);
+    private final CompletableFuture<Void> initWaiter = new CompletableFuture<>();
 
     /**
      * Get all the bundles that are owned by this broker.
@@ -396,8 +395,8 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
             pulsar.runWhenReadyForIncomingRequests(() -> {
                 try {
                     this.serviceUnitStateChannel.start();
-                } catch (PulsarServerException e) {
-                    throw FutureUtil.wrapToCompletionException(e);
+                } catch (Exception e) {
+                    failStarting(e);
                 }
             });
             this.antiAffinityGroupPolicyHelper =
@@ -431,54 +430,72 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
                     new TopBundleLoadDataReporter(pulsar, brokerRegistry.getBrokerId(), topBundlesLoadDataStore);
             this.serviceUnitStateChannel.listen(brokerLoadDataReporter);
             this.serviceUnitStateChannel.listen(topBundleLoadDataReporter);
-            var interval = conf.getLoadBalancerReportUpdateMinIntervalMillis();
-            this.brokerLoadDataReportTask = this.pulsar.getLoadManagerExecutor()
-                    .scheduleAtFixedRate(() -> {
-                                try {
-                                    brokerLoadDataReporter.reportAsync(false);
-                                    // TODO: update broker load metrics using getLocalData
-                                } catch (Throwable e) {
-                                    log.error("Failed to run the broker load manager executor job.", e);
-                                }
-                            },
-                            interval,
-                            interval, TimeUnit.MILLISECONDS);
-
-            this.topBundlesLoadDataReportTask = this.pulsar.getLoadManagerExecutor()
-                    .scheduleAtFixedRate(() -> {
-                                try {
-                                    // TODO: consider excluding the bundles that are in the process of split.
-                                    topBundleLoadDataReporter.reportAsync(false);
-                                } catch (Throwable e) {
-                                    log.error("Failed to run the top bundles load manager executor job.", e);
-                                }
-                            },
-                            interval,
-                            interval, TimeUnit.MILLISECONDS);
-
-            this.monitorTask = this.pulsar.getLoadManagerExecutor()
-                    .scheduleAtFixedRate(() -> {
-                                monitor();
-                            },
-                            MONITOR_INTERVAL_IN_MILLIS,
-                            MONITOR_INTERVAL_IN_MILLIS, TimeUnit.MILLISECONDS);
 
             this.unloadScheduler = new UnloadScheduler(
                     pulsar, pulsar.getLoadManagerExecutor(), unloadManager, context,
                     serviceUnitStateChannel, unloadCounter, unloadMetrics);
             this.splitScheduler = new SplitScheduler(
                     pulsar, serviceUnitStateChannel, splitManager, splitCounter, splitMetrics, context);
-            this.splitScheduler.start();
-            pulsar.runWhenReadyForIncomingRequests(() -> this.initWaiter.countDown());
-            this.started = true;
-            log.info("Started load manager.");
+
+            pulsar.runWhenReadyForIncomingRequests(() -> {
+                try {
+                    var interval = conf.getLoadBalancerReportUpdateMinIntervalMillis();
+
+                    this.brokerLoadDataReportTask = this.pulsar.getLoadManagerExecutor()
+                            .scheduleAtFixedRate(() -> {
+                                        try {
+                                            brokerLoadDataReporter.reportAsync(false);
+                                            // TODO: update broker load metrics using getLocalData
+                                        } catch (Throwable e) {
+                                            log.error("Failed to run the broker load manager executor job.", e);
+                                        }
+                                    },
+                                    interval,
+                                    interval, TimeUnit.MILLISECONDS);
+
+                    this.topBundlesLoadDataReportTask = this.pulsar.getLoadManagerExecutor()
+                            .scheduleAtFixedRate(() -> {
+                                        try {
+                                            // TODO: consider excluding the bundles that are in the process of split.
+                                            topBundleLoadDataReporter.reportAsync(false);
+                                        } catch (Throwable e) {
+                                            log.error("Failed to run the top bundles load manager executor job.", e);
+                                        }
+                                    },
+                                    interval,
+                                    interval, TimeUnit.MILLISECONDS);
+
+                    this.monitorTask = this.pulsar.getLoadManagerExecutor()
+                            .scheduleAtFixedRate(() -> {
+                                        monitor();
+                                    },
+                                    MONITOR_INTERVAL_IN_MILLIS,
+                                    MONITOR_INTERVAL_IN_MILLIS, TimeUnit.MILLISECONDS);
+
+                    this.splitScheduler.start();
+                    this.initWaiter.complete(null);
+                    this.started = true;
+                    log.info("Started load manager.");
+                } catch (Exception ex) {
+                    failStarting(ex);
+                }
+            });
         } catch (Exception ex) {
-            log.error("Failed to start the extensible load balance and close broker registry {}.",
-                    this.brokerRegistry, ex);
-            if (this.brokerRegistry != null) {
+            failStarting(ex);
+        }
+    }
+
+    private void failStarting(Exception ex) {
+        log.error("Failed to start the extensible load balance and close broker registry {}.",
+                this.brokerRegistry, ex);
+        if (this.brokerRegistry != null) {
+            try {
                 brokerRegistry.close();
+            } catch (PulsarServerException e) {
+                // ignore
             }
         }
+        initWaiter.completeExceptionally(ex);
     }
 
     @Override
@@ -824,7 +841,7 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
         boolean becameFollower = false;
         while (!Thread.currentThread().isInterrupted()) {
             try {
-                initWaiter.await();
+                initWaiter.get();
                 if (!serviceUnitStateChannel.isChannelOwner()) {
                     becameFollower = true;
                     break;
@@ -874,7 +891,7 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
         boolean becameLeader = false;
         while (!Thread.currentThread().isInterrupted()) {
             try {
-                initWaiter.await();
+                initWaiter.get();
                 if (serviceUnitStateChannel.isChannelOwner()) {
                     becameLeader = true;
                     break;
@@ -944,7 +961,7 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
     @VisibleForTesting
     protected void monitor() {
         try {
-            initWaiter.await();
+            initWaiter.get();
 
             // Monitor role
             // Periodically check the role in case ZK watcher fails.

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java
@@ -376,12 +376,14 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
                     pulsar.getCoordinationService(), pulsar.getBrokerId(),
                     pulsar.getSafeWebServiceAddress(), ELECTION_ROOT,
                     state -> {
-                        pulsar.getLoadManagerExecutor().execute(() -> {
-                            if (state == LeaderElectionState.Leading) {
-                                playLeader();
-                            } else {
-                                playFollower();
-                            }
+                        pulsar.runWhenReadyForIncomingRequests(() -> {
+                            pulsar.getLoadManagerExecutor().execute(() -> {
+                                if (state == LeaderElectionState.Leading) {
+                                    playLeader();
+                                } else {
+                                    playFollower();
+                                }
+                            });
                         });
                     });
             this.serviceUnitStateChannel = new ServiceUnitStateChannelImpl(pulsar);
@@ -391,7 +393,13 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
             this.serviceUnitStateChannel.listen(unloadManager);
             this.serviceUnitStateChannel.listen(splitManager);
             this.leaderElectionService.start();
-            this.serviceUnitStateChannel.start();
+            pulsar.runWhenReadyForIncomingRequests(() -> {
+                try {
+                    this.serviceUnitStateChannel.start();
+                } catch (PulsarServerException e) {
+                    throw FutureUtil.wrapToCompletionException(e);
+                }
+            });
             this.antiAffinityGroupPolicyHelper =
                     new AntiAffinityGroupPolicyHelper(pulsar, serviceUnitStateChannel);
             antiAffinityGroupPolicyHelper.listenFailureDomainUpdate();
@@ -461,7 +469,7 @@ public class ExtensibleLoadManagerImpl implements ExtensibleLoadManager, BrokerS
             this.splitScheduler = new SplitScheduler(
                     pulsar, serviceUnitStateChannel, splitManager, splitCounter, splitMetrics, context);
             this.splitScheduler.start();
-            this.initWaiter.countDown();
+            pulsar.runWhenReadyForIncomingRequests(() -> this.initWaiter.countDown());
             this.started = true;
             log.info("Started load manager.");
         } catch (Exception ex) {

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java
@@ -1335,7 +1335,9 @@ public class NamespaceService implements AutoCloseable {
                 bundleOwnershipListeners.add(listener);
             }
         }
-        getOwnedServiceUnits().forEach(bundle -> notifyNamespaceBundleOwnershipListener(bundle, listeners));
+        pulsar.runWhenReadyForIncomingRequests(() -> {
+            getOwnedServiceUnits().forEach(bundle -> notifyNamespaceBundleOwnershipListener(bundle, listeners));
+        });
     }
 
     public void addNamespaceBundleSplitListener(NamespaceBundleSplitListener... listeners) {

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java
@@ -19,6 +19,7 @@
 package org.apache.pulsar.broker.service;
 
 import com.google.common.annotations.VisibleForTesting;
+import io.netty.channel.ChannelHandler;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.socket.SocketChannel;
 import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
@@ -104,6 +105,9 @@ public class PulsarChannelInitializer extends ChannelInitializer<SocketChannel> 
 
     @Override
     protected void initChannel(SocketChannel ch) throws Exception {
+        // disable auto read explicitly so that requests aren't served until auto read is enabled
+        // ServerCnx must enable auto read in channelActive after PulsarService is ready to accept incoming requests
+        ch.config().setAutoRead(false);
         ch.pipeline().addLast("consolidation", new FlushConsolidationHandler(1024, true));
         if (this.enableTls) {
             if (this.tlsEnabledWithKeyStore) {
@@ -128,7 +132,8 @@ public class PulsarChannelInitializer extends ChannelInitializer<SocketChannel> 
         // ServerCnx ends up reading higher number of messages and broker can not throttle the messages by disabling
         // auto-read.
         ch.pipeline().addLast("flowController", new FlowControlHandler());
-        ServerCnx cnx = newServerCnx(pulsar, listenerName);
+        // using "ChannelHandler" type to workaround an IntelliJ bug that shows a false positive error
+        ChannelHandler cnx = newServerCnx(pulsar, listenerName);
         ch.pipeline().addLast("handler", cnx);
     }
 

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
@@ -34,7 +34,6 @@ import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Strings;
 import com.google.re2j.Pattern;
 import io.netty.buffer.ByteBuf;
-import io.netty.channel.ChannelConfig;
 import io.netty.channel.ChannelHandler;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelOption;
@@ -370,29 +369,10 @@ public class ServerCnx extends PulsarHandler implements TransportCnx {
         this.commandSender = new PulsarCommandSenderImpl(brokerInterceptor, this);
         this.service.getPulsarStats().recordConnectionCreate();
         cnxsPerThread.get().add(this);
-        // auto read has been disabled by default, handle enabling it
-        handleEnablingAutoRead(ctx);
-    }
-
-    private void handleEnablingAutoRead(ChannelHandlerContext ctx) {
-        ChannelConfig config = ctx.channel().config();
-
-        // serve local requests immediately since some services connect to the broker before it
-        // has been started. This is the case for ExtensibleLoadManager's ServiceUnitStateChannel
-        if (ExtensibleLoadManagerImpl.isLoadManagerExtensionEnabled(service.getPulsar())) {
-            if (ctx.channel().remoteAddress() instanceof InetSocketAddress remote) {
-                if (remote.getAddress() != null && remote.getAddress().isLoopbackAddress()) {
-                    config.setAutoRead(true);
-                }
-            }
-        }
-
-        if (!config.isAutoRead()) {
-            service.getPulsar().runWhenReadyForIncomingRequests(() -> {
-                // enable auto read after PulsarService is ready to accept incoming requests
-                config.setAutoRead(true);
-            });
-        }
+        service.getPulsar().runWhenReadyForIncomingRequests(() -> {
+            // enable auto read after PulsarService is ready to accept incoming requests
+            ctx.channel().config().setAutoRead(true);
+        });
     }
 
     @Override

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
@@ -369,6 +369,10 @@ public class ServerCnx extends PulsarHandler implements TransportCnx {
         this.commandSender = new PulsarCommandSenderImpl(brokerInterceptor, this);
         this.service.getPulsarStats().recordConnectionCreate();
         cnxsPerThread.get().add(this);
+        service.getPulsar().runWhenReadyForIncomingRequests(() -> {
+            // enable auto read after PulsarService is ready to accept incoming requests
+            ctx.channel().config().setAutoRead(true);
+        });
     }
 
     @Override

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
@@ -34,6 +34,7 @@ import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Strings;
 import com.google.re2j.Pattern;
 import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelConfig;
 import io.netty.channel.ChannelHandler;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelOption;
@@ -369,10 +370,29 @@ public class ServerCnx extends PulsarHandler implements TransportCnx {
         this.commandSender = new PulsarCommandSenderImpl(brokerInterceptor, this);
         this.service.getPulsarStats().recordConnectionCreate();
         cnxsPerThread.get().add(this);
-        service.getPulsar().runWhenReadyForIncomingRequests(() -> {
-            // enable auto read after PulsarService is ready to accept incoming requests
-            ctx.channel().config().setAutoRead(true);
-        });
+        // auto read has been disabled by default, handle enabling it
+        handleEnablingAutoRead(ctx);
+    }
+
+    private void handleEnablingAutoRead(ChannelHandlerContext ctx) {
+        ChannelConfig config = ctx.channel().config();
+
+        // serve local requests immediately since some services connect to the broker before it
+        // has been started. This is the case for ExtensibleLoadManager's ServiceUnitStateChannel
+        if (ExtensibleLoadManagerImpl.isLoadManagerExtensionEnabled(service.getPulsar())) {
+            if (ctx.channel().remoteAddress() instanceof InetSocketAddress remote) {
+                if (remote.getAddress() != null && remote.getAddress().isLoopbackAddress()) {
+                    config.setAutoRead(true);
+                }
+            }
+        }
+
+        if (!config.isAutoRead()) {
+            service.getPulsar().runWhenReadyForIncomingRequests(() -> {
+                // enable auto read after PulsarService is ready to accept incoming requests
+                config.setAutoRead(true);
+            });
+        }
     }
 
     @Override

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/WebService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/WebService.java
@@ -21,7 +21,6 @@ package org.apache.pulsar.broker.web;
 import io.prometheus.client.CollectorRegistry;
 import io.prometheus.client.jetty.JettyStatisticsCollector;
 import java.io.IOException;
-import java.net.InetSocketAddress;
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
@@ -40,7 +39,6 @@ import lombok.Getter;
 import org.apache.pulsar.broker.PulsarServerException;
 import org.apache.pulsar.broker.PulsarService;
 import org.apache.pulsar.broker.ServiceConfiguration;
-import org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerImpl;
 import org.apache.pulsar.jetty.tls.JettySslContextFactory;
 import org.eclipse.jetty.server.ConnectionFactory;
 import org.eclipse.jetty.server.ConnectionLimit;
@@ -49,7 +47,6 @@ import org.eclipse.jetty.server.Handler;
 import org.eclipse.jetty.server.HttpConfiguration;
 import org.eclipse.jetty.server.HttpConnectionFactory;
 import org.eclipse.jetty.server.ProxyConnectionFactory;
-import org.eclipse.jetty.server.Request;
 import org.eclipse.jetty.server.RequestLog;
 import org.eclipse.jetty.server.SecureRequestCustomizer;
 import org.eclipse.jetty.server.Server;
@@ -317,17 +314,6 @@ public class WebService implements AutoCloseable {
             @Override
             public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
                     throws IOException, ServletException {
-                if (ExtensibleLoadManagerImpl.isLoadManagerExtensionEnabled(pulsarService)) {
-                    if (request instanceof Request jettyRequest) {
-                        InetSocketAddress remote = jettyRequest.getHttpChannel().getRemoteAddress();
-                        if (remote.getAddress() != null && remote.getAddress().isLoopbackAddress()) {
-                            // skip waiting for local requests when ExtensibleLoadManagerImpl is used
-                            // It connects to the broker before it has been started
-                            chain.doFilter(request, response);
-                            return;
-                        }
-                    }
-                }
                 try {
                     // Wait until the PulsarService is ready to serve incoming requests
                     pulsarService.waitUntilReadyForIncomingRequests();

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/WebService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/WebService.java
@@ -20,12 +20,21 @@ package org.apache.pulsar.broker.web;
 
 import io.prometheus.client.CollectorRegistry;
 import io.prometheus.client.jetty.JettyStatisticsCollector;
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.ExecutionException;
 import javax.servlet.DispatcherType;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
 import lombok.Getter;
 import org.apache.pulsar.broker.PulsarServerException;
 import org.apache.pulsar.broker.PulsarService;
@@ -232,6 +241,7 @@ public class WebService implements AutoCloseable {
         private final FilterHolder authenticationFilterHolder;
         FilterInitializer(PulsarService pulsarService) {
             ServiceConfiguration config = pulsarService.getConfiguration();
+
             if (config.getMaxConcurrentHttpRequests() > 0) {
                 FilterHolder filterHolder = new FilterHolder(QoSFilter.class);
                 filterHolder.setInitParameter("maxRequests", String.valueOf(config.getMaxConcurrentHttpRequests()));
@@ -242,6 +252,10 @@ public class WebService implements AutoCloseable {
                 filterHolders.add(new FilterHolder(
                         new RateLimitingFilter(config.getHttpRequestsMaxPerSecond())));
             }
+
+            // wait until the PulsarService is ready to serve incoming requests
+            filterHolders.add(
+                    new FilterHolder(new WaitUntilPulsarServiceIsReadyForIncomingRequestsFilter(pulsarService)));
 
             boolean brokerInterceptorEnabled = pulsarService.getBrokerInterceptor() != null;
             if (brokerInterceptorEnabled) {
@@ -284,6 +298,42 @@ public class WebService implements AutoCloseable {
             }
         }
 
+        // Filter that waits until the PulsarService is ready to serve incoming requests
+        private static class WaitUntilPulsarServiceIsReadyForIncomingRequestsFilter implements Filter {
+            private final PulsarService pulsarService;
+
+            public WaitUntilPulsarServiceIsReadyForIncomingRequestsFilter(PulsarService pulsarService) {
+                this.pulsarService = pulsarService;
+            }
+
+            @Override
+            public void init(FilterConfig filterConfig) throws ServletException {
+
+            }
+
+            @Override
+            public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+                    throws IOException, ServletException {
+                try {
+                    // Wait until the PulsarService is ready to serve incoming requests
+                    pulsarService.waitUntilReadyForIncomingRequests();
+                } catch (ExecutionException e) {
+                    ((HttpServletResponse) response).sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+                            "PulsarService failed to start.");
+                    return;
+                } catch (InterruptedException e) {
+                    ((HttpServletResponse) response).sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+                            "PulsarService is not ready.");
+                    return;
+                }
+                chain.doFilter(request, response);
+            }
+
+            @Override
+            public void destroy() {
+
+            }
+        }
     }
 
     public void addServlet(String path, ServletHolder servletHolder, boolean requiresAuthentication,

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/WebService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/WebService.java
@@ -21,6 +21,7 @@ package org.apache.pulsar.broker.web;
 import io.prometheus.client.CollectorRegistry;
 import io.prometheus.client.jetty.JettyStatisticsCollector;
 import java.io.IOException;
+import java.net.InetSocketAddress;
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
@@ -39,6 +40,7 @@ import lombok.Getter;
 import org.apache.pulsar.broker.PulsarServerException;
 import org.apache.pulsar.broker.PulsarService;
 import org.apache.pulsar.broker.ServiceConfiguration;
+import org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerImpl;
 import org.apache.pulsar.jetty.tls.JettySslContextFactory;
 import org.eclipse.jetty.server.ConnectionFactory;
 import org.eclipse.jetty.server.ConnectionLimit;
@@ -47,6 +49,7 @@ import org.eclipse.jetty.server.Handler;
 import org.eclipse.jetty.server.HttpConfiguration;
 import org.eclipse.jetty.server.HttpConnectionFactory;
 import org.eclipse.jetty.server.ProxyConnectionFactory;
+import org.eclipse.jetty.server.Request;
 import org.eclipse.jetty.server.RequestLog;
 import org.eclipse.jetty.server.SecureRequestCustomizer;
 import org.eclipse.jetty.server.Server;
@@ -314,6 +317,17 @@ public class WebService implements AutoCloseable {
             @Override
             public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
                     throws IOException, ServletException {
+                if (ExtensibleLoadManagerImpl.isLoadManagerExtensionEnabled(pulsarService)) {
+                    if (request instanceof Request jettyRequest) {
+                        InetSocketAddress remote = jettyRequest.getHttpChannel().getRemoteAddress();
+                        if (remote.getAddress() != null && remote.getAddress().isLoopbackAddress()) {
+                            // skip waiting for local requests when ExtensibleLoadManagerImpl is used
+                            // It connects to the broker before it has been started
+                            chain.doFilter(request, response);
+                            return;
+                        }
+                    }
+                }
                 try {
                     // Wait until the PulsarService is ready to serve incoming requests
                     pulsarService.waitUntilReadyForIncomingRequests();

